api/src/main/java/net/md_5/bungee/api/event/AsyncEvent.java

```diff
@@ -43,39 +42,49 @@ public void postCall()
     /**
      * Register an intent that this plugin will continue to perform work on a
      * background task, and wishes to let the event proceed once the registered
-     * background task has completed.
+     * background task has completed. Multiple intents can be registered by a
+     * plugin, but the plugin must complete the same amount of intents for the
+     * event to proceed.
      *
      * @param plugin the plugin registering this intent
      */
     public void registerIntent(Plugin plugin)
     {
         Preconditions.checkState( !fired.get(), "Event %s has already been fired", this );
-        Preconditions.checkState( !intents.contains( plugin ), "Plugin %s already registered intent for event %s", plugin, this );
 
-        intents.add( plugin );
-        latch.incrementAndGet();
+        if ( !intents.containsKey( plugin ) )
+        {
+            intents.put( plugin, new AtomicInteger( 1 ) );
+            latch.incrementAndGet();
+        } else if ( intents.get( plugin ).getAndIncrement() == 0 )
+        {
+            latch.incrementAndGet();
+        }
     }
 
     /**
-     * Notifies this event that this plugin has done all its required processing
-     * and wishes to let the event proceed.
+     * Notifies this event that this plugin has completed an intent and wishes
+     * to let the event proceed once all intents have been completed.
      *
      * @param plugin a plugin which has an intent registered for this event
      */
     @SuppressWarnings("unchecked")
     public void completeIntent(Plugin plugin)
     {
-        Preconditions.checkState( intents.contains( plugin ), "Plugin %s has not registered intent for event %s", plugin, this );
-        intents.remove( plugin );
-        if ( fired.get() )
+        Preconditions.checkState( intents.containsKey( plugin ) && intents.get( plugin ).get() != 0, "Plugin %s has not registered intents for event %s", plugin, this );
```

Comment:
Fixed that. Originally I was thinking of what if the count went below 0, but now thinking about it more, that would be impossible because a call to registerIntent must have happened before the call to completeIntent.

Reply:
Still better to have the check like that.